crates/polars-json/src/ndjson/deserialize.rs

```diff
@@ -14,27 +16,42 @@ use super::*;
 pub fn deserialize_iter<'a>(
     rows: impl Iterator<Item = &'a str>,
     data_type: ArrowDataType,
-    buf_size: usize,
-    count: usize,
 ) -> PolarsResult<ArrayRef> {
-    let mut buf = String::with_capacity(buf_size + count + 2);
+    let mut arr: Vec<Box<dyn Array>> = Vec::new();
+    let mut buf = String::with_capacity(std::u32::MAX as usize);
     buf.push('[');
+
+    fn _deserializer(s: String, data_type: ArrowDataType) -> PolarsResult<Box<dyn Array>> {
+        let mut buf = s.clone();
+        let slice = unsafe { buf.as_bytes_mut() };
+        let out = simd_json::to_borrowed_value(slice)
+            .map_err(|e| PolarsError::ComputeError(format!("json parsing error: '{e}'").into()))?;
+        Ok(if let BorrowedValue::Array(rows) = out {
+            super::super::json::deserialize::_deserialize(&rows, data_type.clone())
+        } else {
+            unreachable!()
+        })
+    }
+
     for row in rows {
         buf.push_str(row);
-        buf.push(',')
+        buf.push(',');
+
+        if buf.len() + row.len() > std::u32::MAX as usize {
```

Comment:
If the row lengths vary a lot, this can fail. Give some more leeway.

```suggestion
        if buf.len() + row.len() > (std::u32::MAX << 1) as usize {
```

Reply:
@ritchie46 what does this left shift do exactly?